### PR TITLE
IKDT-648 build the typeahead index

### PR DESCRIPTION
**Pull request by dettmoney**

### CODE REVIEW

File: entity/src/main/java/dev/ikm/tinkar/entity/load/LoadEntitiesFromProtobufFile.java [hunk 1]
```diff
@@ -77,6 +84,7 @@ public EntityCountSummary compute() {
         updateMessage("Importing Protobuf Data...");
 
         // Process Protobuf Entry
+        EntityService.get().beginLoadPhase();
```
Comment (stenkarl) [inline entity/src/main/java/dev/ikm/tinkar/entity/load/LoadEntitiesFromProtobufFile.java:87]: Do we also need to call these load lifecycle methods when we load the Spined Array store?

File: integration/src/test/java/dev/ikm/tinkar/integration/search/SearcherIT.java [hunk 2]
```diff
@@ -160,7 +167,87 @@ public void searchFromDescendantsOfConceptWithCustomCalculatorIT() throws Except
...
+        PublicId ancestorId = EntityService.get().getEntity(UUID.fromString("f7495b58-6630-3499-a44e-2052b5fcf06c")).get().publicId();
+
+        String userInput = "u";
+        List<LatestVersionSearchResult> allSearchResults = TypeAheadSearch.typeAheadSuggestions(userInput, ancestorId);
+
+        stopwatch.stop();
+    }
+
+    @Test
+    public void typeAheadIndexerTestSearchAndDescendants2() throws Exception {
```
Comment (stenkarl) [inline integration/src/test/java/dev/ikm/tinkar/integration/search/SearcherIT.java:203]: These next few tests don't assert anything. Maybe they are in-progress though.

File: integration/src/test/java/dev/ikm/tinkar/integration/search/SearcherIT.java [hunk 3]
```diff
@@ -160,7 +167,87 @@ public void searchFromDescendantsOfConceptWithCustomCalculatorIT() throws Except
...
+        Stopwatch stopwatch = new Stopwatch();
+
+        String userInput = "u";
+        List<LatestVersionSearchResult> allSearchResults = TypeAheadSearch.typeAheadFuzzySuggestions(userInput, ancestorId);
+
+        stopwatch.stop();
+    }
+
+    @Test
+    public void typeAheadIndexerTestSearchAndDescendantsFuzzy2() throws Exception {
```
Comment (stenkarl) [inline integration/src/test/java/dev/ikm/tinkar/integration/search/SearcherIT.java:234]: I get this exception when running this test FYI:

`java.lang.IllegalStateException: No DataServiceController selected for provider. Tried 0`

File: provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        DirectoryReader reader = DirectoryReader.open(Indexer.indexWriter);
+        LuceneDictionary dict = new LuceneDictionary(reader, TEXT_FIELD_NAME);
+        IndexSearcher indexSearcher = new IndexSearcher(reader);
+        QueryParser parser = new QueryParser("text", Indexer.analyzer());
+        Query query = parser.parse(term);
+        Formatter formatter = new SimpleHTMLFormatter();
+        QueryScorer scorer = new QueryScorer(query);
+        Highlighter highlighter = new Highlighter(formatter, scorer);
+        ScoreDoc[] hits = indexSearcher.search(query, 100).scoreDocs;
+        System.out.println(hits);
```
Comment (stenkarl) [inline provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java:109]: There are a few System.out.printlns in here probably left over from testing.

File: provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            StoredField rcNidField = (StoredField) hitDoc.getField(Indexer.RC_NID);
+            StoredField fieldIndexField = (StoredField) hitDoc.getField(Indexer.FIELD_INDEX);
+            StoredField textField = (StoredField) hitDoc.getField(Indexer.TEXT_FIELD_NAME);
+            String highlightedString = highlighter.getBestFragment(Indexer.analyzer(), Indexer.TEXT_FIELD_NAME, textField.stringValue());
+
+            results[i] = new PrimitiveDataSearchResult(nidField.numericValue().intValue(), rcNidField.numericValue().intValue(),
+                    patternNidField.numericValue().intValue(), fieldIndexField.numericValue().intValue(), hits[i].score, highlightedString);
+            System.out.println(results[i]);
+        }
+        System.out.println(results);
```
Comment (stenkarl) [inline provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java:125]: System.out.printlns.

File: provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        fuzzySuggester.build(dict);
+        LOG.info("TypeAheadSearch fuzzy index build duration: {}", stopwatch.durationString());
+    }
+
+    public static List<String> fuzzySuggest(String term) throws IOException {
+        List<Lookup.LookupResult> lookup = fuzzySuggester.lookup(term, false, 100);
+        return lookup.stream().map(a -> a.key.toString()).collect(Collectors.toList());
+    }
+
+    public static void buildSearchIndex(String term) throws IOException, ParseException, InvalidTokenOffsetsException {
```
Comment (stenkarl) [inline provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java:99]: This is never called.

File: provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/Indexer.java [hunk 7]
```diff
@@ -54,9 +51,9 @@ public class Indexer {
     private static final Logger LOG = LoggerFactory.getLogger(Indexer.class);
     private static final File defaultDataDirectory = new File("target/lucene/");
     private static DirectoryReader indexReader;
-    private static Directory indexDirectory;
-    private static Analyzer analyzer;
-    private static IndexWriter indexWriter;
+    static Directory indexDirectory;
```
Comment (stenkarl) [inline provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/Indexer.java:54]: Could we make these three variables private and expose them with getters? That way we can prevent them from getting reassigned outside this class.

File: provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class TypeAheadSearch {
```
Comment (stenkarl) [inline provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java:55]: Could we make this class a singleton and remove all of the static modifiers? I realize that almost everything in the codebase follows a static pattern, but a revolution can begin with just one class...

File: integration/src/test/java/dev/ikm/tinkar/integration/search/SearcherIT.java (hunk 2 above)
Comment (dettmoney) [reply integration/src/test/java/dev/ikm/tinkar/integration/search/SearcherIT.java:203]: Yes these are in progress and will be addressed in the next typeahead task.

File: provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java (hunk 4 above)
Comment (dettmoney) [reply provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java:109]: these are in progress and will be addressed in the next typeahead task.

File: provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java (hunk 5 above)
Comment (dettmoney) [reply provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java:125]: these are in progress and will be addressed in the next typeahead task.

File: provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java (hunk 6 above)
Comment (dettmoney) [reply provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java:99]: these are in progress and will be addressed in the next typeahead task.

File: provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java (hunk 8 above)
Comment (dettmoney) [reply provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java:55]: Done in next commit

File: provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/Indexer.java (hunk 7 above)
Comment (dettmoney) [reply provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/Indexer.java:54]: Done in next commit

File: entity/src/main/java/dev/ikm/tinkar/entity/load/LoadEntitiesFromProtobufFile.java (hunk 1 above)
Comment (dettmoney) [reply entity/src/main/java/dev/ikm/tinkar/entity/load/LoadEntitiesFromProtobufFile.java:87]: The Spined Array Store looks like it is directly accessed as bytes on disk. As such, the usual method of calling a put API to add an Entity to the database is not used (at least as far as I can tell). The starter-data-builder and snomedct-transformer code can be updated in the future to make use of the lifecycle methods to reduce their execution time.

**Pull request merged**